src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs

```diff
@@ -159,33 +160,34 @@ private IBlockImprovementContext CreateBlockImprovementContext(string payloadId,
                 }
             },
             TaskContinuationOptions.RunContinuationsAsynchronously);
-        blockImprovementContext.ImprovementTask.ContinueWith(async _ =>
+        blockImprovementContext.ImprovementTask.ContinueWith(async b =>
         {
             CancellationToken token = cts.Token;
-            if (token.IsCancellationRequested)
+            do
             {
-                return;
-            }
+                if (token.IsCancellationRequested)
+                {
+                    return;
+                }
 
-            TimeSpan dynamicDelay = CalculateImprovementDelay(startDateTime, startTimestamp);
-            if (dynamicDelay == Timeout.InfiniteTimeSpan)
-            {
-                // If we can't finish before that cutoff, skip the improvement
-                if (_logger.IsTrace) _logger.Trace($"Block for payload {payloadId} with parent {parentHeader.ToString(BlockHeader.Format.FullHashAndNumber)} won't be improved, no more time in slot");
-                return;
-            }
+                TimeSpan dynamicDelay = CalculateImprovementDelay(startDateTime, startTimestamp);
+                if (dynamicDelay == Timeout.InfiniteTimeSpan)
+                {
+                    // If we can't finish before that cutoff, skip the improvement
+                    if (_logger.IsTrace) _logger.Trace($"Block for payload {payloadId} with parent {parentHeader.ToString(BlockHeader.Format.FullHashAndNumber)} won't be improved, no more time in slot");
+                    return;
+                }
 
-            // If we reach here, we still have time for an improvement build (which still responds to cancellation)
-            try
-            {
-                // If the dynamic delay is still positive, await that
-                if (dynamicDelay > TimeSpan.Zero)
+                // If we reach here, we still have time for an improvement build (which still responds to cancellation)
+                try
                 {
-                    if (_logger.IsTrace) _logger.Trace($"Block for payload {payloadId} with parent {parentHeader.ToString(BlockHeader.Format.FullHashAndNumber)} will be improved in {dynamicDelay.TotalMilliseconds}ms");
                     await Task.Delay(dynamicDelay, token);
                 }
-            }
-            catch (OperationCanceledException) { }
+                catch (OperationCanceledException) { }
+
+                // Loop the delay if no new txs have been added, and while not cancelled.
+                // Is no point in rebuilding an identical block.
+            } while (added == TxPool.Metrics.PendingTransactionsAdded);
```

Comment:
Should `TxPool.Metrics.PendingTransactionsAdded` be changed to a field and we should do `Volatile.Read` of it?

Reply:
I thought about this; decided not to

But you are probably right 😄

Reply:
Done